File: examples/java/patterns/idempotency/choose-semantics.java
import software.amazon.lambda.durable.config.StepConfig;
import software.amazon.lambda.durable.config.StepSemantics;
import software.amazon.lambda.durable.retry.RetryStrategies;

// At-least-once (default).
context.step("upsert-user", User.class,
    ctx -> userStore.upsert(input.user()));

// At-most-once with retries disabled.
StepConfig critical = StepConfig.builder()
    .semantics(StepSemantics.AT_MOST_ONCE_PER_RETRY)
    .retryStrategy(RetryStrategies.Presets.NO_RETRY)
    .build();

context.step("charge-payment", Receipt.class,
    ctx -> paymentService.charge(input.amount(), input.cardToken()),
    critical);

File: examples/typescript/patterns/idempotency/choose-semantics.ts
import { StepSemantics } from "@aws/durable-execution-sdk-js";

// At-least-once (default) for a retryable idempotent write.
await context.step("upsert-user", async () => {
  return userStore.upsert(event.user);
});

// At-most-once for a side-effecting call, with retries disabled.
await context.step(
  "charge-payment",
  async () => paymentService.charge(event.amount, event.cardToken),
  {
    semantics: StepSemantics.AtMostOncePerRetry,
    retryStrategy: () => ({ shouldRetry: false }),
  },
);
